src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs

```diff
@@ -88,11 +90,11 @@ public void entity_query()
     }
 
     [Theory]
-    [InlineData("...on User { id }", false)]
-    [InlineData("__typename ...on User { id }", false)]
-    [InlineData("...on User { __typename id }", false)]
-    [InlineData("...on User { ...TypeAndId }", true)]
-    public void result_includes_typename(string selectionSet, bool includeFragment)
+    [InlineData("...on User { id }", false, true)]
+    [InlineData("__typename ...on User { id }", false, false)]
+    [InlineData("...on User { __typename id }", false, false)]
+    [InlineData("...on User { ...TypeAndId }", true, false)]
+    public void result_includes_typename(string selectionSet, bool includeFragment, bool addTypenameRule)
```

Comment:
If an existing GraphQL.NET user requires `__typename` to be automatically injected into the query, we are providing (for GraphQL.NET v8 only) a validation rule which will perform the injection.  And if it is necessary for certain router implementations, we can then elect to remove the `[Obsolete]` marker from the class.